packages/ui-a11y-utils/src/FocusRegion.ts

```diff
@@ -61,8 +61,11 @@ class FocusRegion {
     this._id = uid()
   }
 
-  updateElement(element: Element | Node) {
+  updateElement(element: Element | Node, options?: FocusRegionOptions) {
     this._contextElement = element
+    if (options) {
+      this._options = options
+    }
     if (this._keyboardFocusRegion) {
       this._keyboardFocusRegion.updateElement(element)
     }
```

Comment:
This was the issue:
the `FocusRegion`'s `this._options` are holding the callback references (such as `onDismiss`) for `Dialog`.
For some reason these callback references were not in sync with the actual state of the `Dialog` and were holding the same reference from the first time `Dialog` was mounted in the dom, thus were calling stale functions with stale state values.
This issue for some reason was only an issue for function based components.